File: userspace/sdjagent/src/main/java/com/sysdigcloud/sdjagent/PosixQueue.java
package com.sysdigcloud.sdjagent;

import java.io.IOException;
import java.nio.charset.Charset;
import java.util.logging.Logger;

/**
 * Created by luca on 18/05/16.
 */
public class PosixQueue {
    private static final Logger LOGGER = Logger.getLogger(PosixQueue.class.getName());

    public enum Direction {
        SEND(0), RECEIVE(1);

        private final int value;
        Direction(int value) {
            this.value = value;
        }
    }

    static {
        if (CLibrary.libraryLoaded) {
            if(!setQueueLimits()) {
                LOGGER.warning("Cannot set queue limits");
            }
        }
    }

    private final int fd;
    private final String name;
    private final byte[] msgbuffer;
    public PosixQueue(String name, Direction direction) throws IOException {
        this.name = name;
        this.msgbuffer = new byte[3 << 20]; // 3MiB
        if (CLibrary.libraryLoaded) {
            int res = openQueue(name, direction.value, 1);
            if (res > 0) {
                this.fd = res;
            } else {
                throw new IOException(String.format("Cannot create posix queue %s errno=%d", name, -res));
            }
        } else {
            throw new IOException(String.format("Cannot create posix queue %s (libsdjagentjni not loaded)", name));
        }
    }

    public boolean send(String message) {
        int res = queueSend(fd, message);
        switch(res) {
            case 0:
                break;
            case -1:
                LOGGER.fine(String.format("Cannot send on queue %s, is full", this.name));
                break;
            case -2:
                LOGGER.warning(String.format("Cannot send on queue %s, msg too big", this.name));
                break;
            default:
                LOGGER.warning(String.format("Cannot send on queue %s, errno: %d", this.name, res));
                break;
        }
        return res == 0;
    }

    public String receive(long timeout_s) throws IOException {
        int res = queueReceive(fd, this.msgbuffer, timeout_s);
        if(res > 0) {
            return new String(this.msgbuffer, 0, res, Charset.defaultCharset());
        } else if (res == -1) {
            return null;
        } else {
            throw new IOException(String.format("Unexpected errno=%d from posix queue receive", -res));
        }
    }

    public void close() {
        queueClose(fd);
    }

    private static native boolean setQueueLimits();
    private static native int openQueue(String name, int flags, int maxmsgs);
    private static native int queueSend(int fd, String message);
    private static native int queueReceive(int fd, byte[] msgbuffer, long timeout_s);
    private static native boolean queueClose(int fd);
}

File: userspace/libsanalyzer/src/posix_queue.cpp
//
// Created by Luca Marturana on 30/06/15.
//

#include "posix_queue.h"
#include <sinsp_int.h>
#include <sys/stat.h>
#include <sys/resource.h>
#include <limits.h>

posix_queue::posix_queue(std::string name, direction_t dir, long maxmsgs):
	m_direction(dir),
	m_name(std::move(name))
{
	m_readbuffer = new char[MAX_MSGSIZE];
	ASSERT(name.size() <= NAME_MAX);
	int flags = dir | O_CREAT;
	struct mq_attr queue_attrs = {0};
	if(m_direction == SEND)
	{
		// We need non_blocking mode only for send
		// on receive we use a timeout
		flags |= O_NONBLOCK;
		queue_attrs.mq_flags = O_NONBLOCK;
	}
	queue_attrs.mq_maxmsg = maxmsgs;
	queue_attrs.mq_msgsize = MAX_MSGSIZE;
	queue_attrs.mq_curmsgs = 0;
	m_queue_d = mq_open(m_name.c_str(), flags, S_IRWXU, &queue_attrs);
	if(m_queue_d < 0)
	{
		g_logger.format(sinsp_logger::SEV_ERROR, "Error: Cannot create queue %s, errno: %s", m_name.c_str(), strerror(errno));
	}
}

posix_queue::~posix_queue()
{
	if(m_queue_d > 0)
	{
		mq_close(m_queue_d);
	}
	delete[] m_readbuffer;
}

bool posix_queue::send(const std::string &msg)
{
	if(m_queue_d)
	{
		auto res = mq_send(m_queue_d, msg.c_str(), msg.size(), 0);
		if(res == 0)
		{
			return true;
		}
		else
		{
			switch(errno)
			{
			case EAGAIN:
				g_logger.format(sinsp_logger::SEV_DEBUG, "Debug: Cannot send on queue %s, is full", m_name.c_str());
				break;
			case EMSGSIZE:
				g_logger.format(sinsp_logger::SEV_WARNING, "Warning: Cannot send on queue %s, msg too big size=%u", m_name.c_str(), msg.size());
				break;
			default:
				g_logger.format(sinsp_logger::SEV_WARNING, "Warning: Cannot send on queue %s, errno: %s", m_name.c_str(), strerror(errno));
				break;
			}
			return false;
		}
	}

	g_logger.log("Error: posix_queue[" + m_name + "]: cannot send (no queue)", sinsp_logger::SEV_ERROR);
	return false;
}

std::string posix_queue::receive(uint64_t timeout_s)
{
	if(m_queue_d)
	{
		struct timespec ts = {0};
		uint64_t now = sinsp_utils::get_current_time_ns();
		ts.tv_sec = now / ONE_SECOND_IN_NS + timeout_s;
		unsigned int prio = 0;
		auto res = mq_timedreceive(m_queue_d, m_readbuffer, MAX_MSGSIZE, &prio, &ts);
		if(res >= 0)
		{
			return std::string(m_readbuffer, res);
		} else if (errno == ETIMEDOUT || errno == EINTR) {
			return "";
		} else {
			g_logger.format(sinsp_logger::SEV_ERROR, "Unexpected error on posix queue receive: %s", strerror(errno));
			if(timeout_s > 0)
			{
				// At this point the application may go to infinite loop if it relies
				// on the timeout provided, eg:
				// while(true)
				// {
				//   auto msg = receive(1)
				//   do stuff...
				// }
				// in this case is better to crash
				// otherwise if timeout=0 like our dragent. let's keep it running
				// as posix queue healthness is not vital
				throw sinsp_exception("Unexpected error on posix queue receive");
			}
			return "";
		}
	}

	g_logger.log("Error: posix_queue[" + m_name + "]: cannot receive (no queue)", sinsp_logger::SEV_ERROR);
	return "";

}

bool posix_queue::remove(const std::string &name)
{
	return mq_unlink(name.c_str()) == 0;
}
